athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,280 @@
+            }
+            LOGGER.info("Examining folder {}", folderPath);
+            List<AbstractMap.SimpleImmutableEntry<String, String>> partitions = PartitionUtil.getPartitionColumnData(table, folderPath);
+            if (!partitions.isEmpty() && checkPartitionWithConstrains(partitions, expressions)) {
+                partitionFolders.add(partitions);
+                LOGGER.info("Folder {} is selected", folderPath);
+            }
+        }
+        return partitionFolders;
+    }
+
+    /**
+     * Retrieves a first of file has non-zero size
+     *
+     * @param bucket Name of the bucket
+     * @param prefix Prefix (aka, folder in Storage service) of the bucket from where this method with retrieve files
+     * @return A single file name under the prefix
+     */
+    protected Optional<String> getStorageFiles(String bucket, String prefix)
+    {
+        LOGGER.info("Listing nested files for prefix {} under the bucket {}", prefix, bucket);
+        Page<Blob> blobPage = storage.list(bucket, prefix(prefix.substring(1)));
+        for (Blob blob : blobPage.iterateAll()) {
+            if (isBlobFile(blob)) {
+               return Optional.of(blob.getName());
+            }
+        }
+        return Optional.empty();
+    }
+
+    // helpers
+    /**
+     * check whether it is file, It may return folder also
+     */
+    private boolean isBlobFile(Blob blob)
+    {
+        return blob.getSize() > 0;
+    }
+
+    @VisibleForTesting
+    protected Schema getFileSchema(String bucketName, String path, FileFormat format, BufferAllocator allocator)
+    {
+        requireNonNull(bucketName, "bucketName was null");
+        requireNonNull(path, "fileName was null");
+        LOGGER.info("Retrieving field schema from file {}, under the bucket {}", path, bucketName);
+        String uri = createUri(bucketName, path);
+        DatasetFactory factory = new FileSystemDatasetFactory(allocator,
+                NativeMemoryPool.getDefault(), format, uri);
+        // inspect schema
+        return factory.inspect();
+    }
+
+    private boolean checkPartitionWithConstrains(List<AbstractMap.SimpleImmutableEntry<String, String>> partitionList, List<AbstractExpression> expressions)
+    {
+        if (expressions.isEmpty()) {
+            return true;
+        }
+        for (AbstractMap.SimpleImmutableEntry<String, String> partition : partitionList) {
+            List<AbstractExpression> expressionList = expressions.stream()
+                    .filter(expr -> expr.columnName.equalsIgnoreCase(partition.getKey()))
+                    .collect(Collectors.toList());
+            for (AbstractExpression expression : expressionList) {
+                LOGGER.debug("Evaluating field value {} against the expression {}", partition, expressions);
+                if (!expression.apply(partition.getValue())) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
```

Comment:
I don't remember but I think I mentioned that this could be rewritten to be more efficient and easier to reason about.
I kept the logic the same, but please double check to make sure that the logic is what you actually want.
Currently in english, this is saying:
"Return failure if we find any partition entries have any expression failures.
No expressions is considered a success."

```
private boolean checkPartitionWithConstrains(List<AbstractMap.SimpleImmutableEntry<String, String>> partitionList, List<AbstractExpression> expressions)
{
    if (expressions.isEmpty() {
        return true;
    }

    // NOTE: expressionMap should actually be created earlier because checkPartitionWithConstrains is being called in a loop where expressions doesn't change per iteration.
    // Put this expressionMap creation after this line:
    // getPartitionFolders():
    // 145: LOGGER.info("Expressions for the request of {}.{} is \n{}", tableInfo.getSchemaName(), tableInfo.getTableName(), expressions);
    // And then just pass in the expressionMap as a parameter to this method instead of creating the expressionMap every iteration.

    Map<String, List<AbstractExpression>> expressionMap = expressions.stream()
        .collect(groupingBy(
            expr -> expr.columnName,
            () -> new TreeMap(String.CASE_INSENSITIVE_ORDER),
            toCollection(ArrayList::new)));

    boolean expressionFailureExists = partitionList.stream().anyMatch(partitionEntry ->
        (expressionMap.getOrDefault(partitionEntry.getKey(), List.of()).stream()
            .anyMatch(expr -> {
                boolean result = expr.apply(partitionEntry.getValue());
                LOGGER.debug("Partition entry: {}, expression: {}, result: {}", partitionEntry, expr, result);
                return !result;
             }))
    );

    return !expressionFailureExists;
}
```

Reply:
I just fixed this for you here: https://github.com/trianz-azam/aws-athena-query-federation/commit/ab6d61fbecbff187ae7aec454c4eb719c7da5009